service/history/health_signal_aggregator.go

```diff
@@ -0,0 +1,198 @@
+func NewHealthSignalAggregatorImpl(
+	aggregationEnabled bool,
+	windowSize time.Duration,
+	maxBufferSize int,
+	metricsHandler metrics.Handler,
+	perMethodRPSWarnLimit dynamicconfig.IntPropertyFn,
+	perMethodErrorWarnLimit dynamicconfig.FloatPropertyFn,
+	logger log.Logger,
+) *HealthSignalAggregatorImpl {
+	ret := &HealthSignalAggregatorImpl{
+		status:                  common.DaemonStatusInitialized,
+		shutdownCh:              make(chan struct{}),
+		requestCounts:           make(map[string]int64),
+		metricsHandler:          metricsHandler,
+		emitMetricsTimer:        time.NewTicker(emitMetricsInterval),
+		perMethodRPSWarnLimit:   perMethodRPSWarnLimit,
+		perMethodErrorWarnLimit: perMethodErrorWarnLimit,
+		logger:                  logger,
+		aggregationEnabled:      aggregationEnabled,
+	}
+
+	if aggregationEnabled {
+		ret.latencyAverage = aggregate.NewMovingWindowAvgImpl(windowSize, maxBufferSize)
+		ret.errorRatio = aggregate.NewMovingWindowAvgImpl(windowSize, maxBufferSize)
+	} else {
+		ret.latencyAverage = aggregate.NoopMovingWindowAverage
+		ret.errorRatio = aggregate.NoopMovingWindowAverage
+	}
+
+	return ret
+}
+
+func (s *HealthSignalAggregatorImpl) Start() {
+	if !atomic.CompareAndSwapInt32(&s.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+	go s.emitMetricsLoop()
+}
+
+func (s *HealthSignalAggregatorImpl) Stop() {
+	if !atomic.CompareAndSwapInt32(&s.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+	close(s.shutdownCh)
+	s.emitMetricsTimer.Stop()
+}
+
+func (s *HealthSignalAggregatorImpl) Record(method string, latency time.Duration, err error) {
+	if s.aggregationEnabled {
+		s.latencyAverage.Record(latency.Milliseconds())
+
+		if isUnhealthyError(err) {
+			s.errorRatio.Record(1)
+		} else {
+			s.errorRatio.Record(0)
+		}
+	}
+
+	s.incrementMethodRequestCount(method)
+}
+
+func (s *HealthSignalAggregatorImpl) AverageLatency() float64 {
+	return s.latencyAverage.Average()
+}
+
+func (s *HealthSignalAggregatorImpl) ErrorRatio() float64 {
+	return s.errorRatio.Average()
+}
+
+func (s *HealthSignalAggregatorImpl) incrementMethodRequestCount(method string) {
+	s.requestsLock.Lock()
+	defer s.requestsLock.Unlock()
+	s.requestCounts[method]++
+}
+
+func (s *HealthSignalAggregatorImpl) emitMetricsLoop() {
+	for {
+		select {
+		case <-s.shutdownCh:
+			return
+		case <-s.emitMetricsTimer.C:
+			s.requestsLock.Lock()
+			requestCounts := s.requestCounts
+			s.requestCounts = make(map[string]int64, len(requestCounts))
+			s.requestsLock.Unlock()
+
+			for method, count := range requestCounts {
+				methodRPS := int64(float64(count) / emitMetricsInterval.Seconds())
+				s.metricsHandler.Histogram(HistoryServiceRPS.Name(), HistoryServiceRPS.Unit()).Record(methodRPS)
+				if methodRPS > int64(s.perMethodRPSWarnLimit()) {
+					s.logger.Warn("Per method RPS warn limit exceeded",
+						tag.NewStringTag("method", method),
+						tag.NewInt64("rps", methodRPS))
+				}
```

Comment:
what is the reason to emit those metrics?

Reply:
Discussed with @prathyushpv and there's no need as we're already doing this. Moved to noop.